packages/statemanager/src/cache.ts

```diff
@@ -1,192 +1,366 @@
 
-export type getCb = (address: Address) => Promise<Account | undefined>
+const LRU = require('lru-cache')
+
+export type getCb = (address: Address) => Promise<Buffer | undefined>
 export type putCb = (keyBuf: Buffer, accountRlp: Buffer) => Promise<void>
 export type deleteCb = (keyBuf: Buffer) => Promise<void>
 
 export interface CacheOpts {
+  size: number
+  type: CacheType
   getCb: getCb
   putCb: putCb
   deleteCb: deleteCb
 }
 
+export enum CacheType {
+  LRU = 'lru',
+  ORDERED_MAP = 'ordered_map',
+}
+
+/**
+ * account: undefined
+ *
+ * Account is known to not exist in the trie
+ */
+type CacheElement = {
+  accountRLP: Buffer | undefined
+}
+
+/**
+ * Diff cache collecting the state of the cache
+ * at the beginning of checkpoint height
+ * (respectively: before a first modification)
+ *
+ * If the whole cache element is undefined (in contrast
+ * to the account), the element didn't exist in the cache
+ * before.
+ */
+type DiffCache = OrderedMap<string, CacheElement | undefined>[]
+
 /**
  * @ignore
  */
 export class Cache {
-  _cache: OrderedMap<any, any>
-  _cacheEnd: OrderedMapIterator<any, any>
-  _checkpoints: any[]
+  _debug: Debugger
+
+  _lruCache: LRUCache<string, CacheElement> | undefined
+  _orderedMapCache: OrderedMap<string, CacheElement> | undefined
+
+  _diffCache: DiffCache = []
+  _checkpoints = 0
 
   _getCb: getCb
   _putCb: putCb
   _deleteCb: deleteCb
 
+  _stats = {
+    cache: {
+      size: 0,
+      reads: 0,
+      hits: 0,
+      writes: 0,
+      dels: 0,
+    },
+    trie: {
+      reads: 0,
+      writes: 0,
+      dels: 0,
+    },
+  }
+
   constructor(opts: CacheOpts) {
-    this._cache = new OrderedMap()
-    this._cacheEnd = this._cache.end()
+    this._debug = createDebugLogger('statemanager:cache')
+
+    if (opts.type === CacheType.LRU) {
+      this._lruCache = new LRU({
+        max: opts.size,
+        updateAgeOnGet: true,
+      })
+    } else {
+      this._orderedMapCache = new OrderedMap()
+    }
+
+    this._diffCache.push(new OrderedMap())
+
     this._getCb = opts.getCb
     this._putCb = opts.putCb
     this._deleteCb = opts.deleteCb
-    this._checkpoints = []
+  }
+
+  _saveCachePreState(addressHex: string) {
+    const it = this._diffCache[this._checkpoints].find(addressHex)
+    if (it.equals(this._diffCache[this._checkpoints].end())) {
+      let oldElem
+      if (this._lruCache) {
+        oldElem = this._lruCache!.get(addressHex)
+      } else {
+        oldElem = this._orderedMapCache!.getElementByKey(addressHex)
+      }
+      this._debug(
+        `Save pre cache state ${
+          oldElem?.accountRLP ? 'as exists' : 'as non-existent'
+        } for account ${addressHex}`
+      )
+      this._diffCache[this._checkpoints].setElement(addressHex, oldElem)
+    }
   }
 
   /**
    * Puts account to cache under its address.
-   * @param key - Address of account
+   * @param key - Address of account or undefined if account doesn't exist in the trie
```

Comment:
Can we rewrite the docs to match the existing function spec?  You now have `address` and `account` where `account` can be either an account or `undefined` so not sure if the above docs still match.

Reply:
Generally did some additional round through the code docs.